File: Backend/backend-api/src/ragflow_client.ts
/**
 * ragflow_client.ts — RAGFlow API Client , connects this backend to flask
 *
 * Provides type-safe functions for interacting with Flask API endpoints directly.
 * Handles:
 * - Dataset management and resolution
 * - Document upload, parsing, and deletion
 * - Query and retrieval operations
 * - Chat completions with optional streaming
 * - Session management
 * - Resource recommendations
 *
 * All calls go directly to Flask AI proxy
 */

import fetch, { HeadersInit } from 'node-fetch';
import FormData from 'form-data';
import { Readable } from 'stream';
import dotenv from 'dotenv';
import { fileURLToPath } from 'url';
import path from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

dotenv.config({
  path: path.resolve(__dirname, '../../..', '.env'),
});

const AI_SERVICE_URL = process.env.AI_SERVICE_URL || 'http://localhost:5001';
const RAGFLOW_API_KEY = process.env.RAGFLOW_API_KEY || '';
const API_BASE = `${AI_SERVICE_URL}/api/ragflow` || 'http://localhost:5001/api/ragflow';

type Json = Record<string, any>;

// ─────────────────────────────────────────────────────────────────────────────
// Type Definitions
// ─────────────────────────────────────────────────────────────────────────────

export interface LessonPlan {
  title: string;
  class_name: string;
  subject: string;
  topic: string;
  duration_minutes: number;
  learning_objectives: string[];
  materials_needed: string[];
  sections: {
    name: string;
    duration_minutes: number;
    content: string;
    activities: string[];
  }[];
  differentiation: {
    struggling: string;
    advanced: string;
  };
  zero_resource_tips: string[];
}

export interface RAGResource {
  name: string;
  url: string;
  state?: string;
  language: string[];
  description: string;
  is_exemplary: boolean;
  tags: string[];
}

export interface RAGDataset {
  id: string;
  name: string;
  description?: string;
  chunk_method?: string;
  [key: string]: any;
}

export interface RAGDocument {
  id?: string;
  document_id?: string;
  name?: string;
  filename?: string;
  status?: string;
  [key: string]: any;
}

export interface RAGChunk {
  id?: string;
  content?: string;
  score?: number;
  similarity?: number;
  document_id?: string;
  [key: string]: any;
}

/**
 * DatasetRef provides a stable reference to a dataset.
 * Maintains both the RAGFlow UUID and human-readable name.
 */
export interface DatasetRef {
  appKey: string;
  ragflowId: string;
  ragflowName: string;
}

// ─────────────────────────────────────────────────────────────────────────────
// Helper Functions
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Get headers for RAGFlow API calls.
 */
function getHeaders(customHeaders: Record<string, string> = {}): HeadersInit {
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    ...customHeaders,
  };
  if (RAGFLOW_API_KEY) {
    headers['Authorization'] = `Bearer ${RAGFLOW_API_KEY}`;
  }
  return headers;
}

/**
 * Parse JSON response, handling both JSON and plain text responses.
 * FIXED: Removed explicit Response type to avoid node-fetch vs undici conflict.
 */
async function parseJson<T = any>(response: any): Promise<T> {
  const text = await response.text();
  let data: any = {};

  try {
    data = text ? JSON.parse(text) : {};
  } catch {
    data = { raw: text };
  }

  if (!response.ok) {
    throw new Error(data?.error || data?.detail || `Request failed with ${response.status}`);
  }

  return data as T;
}

/**
 * Execute GET request and parse JSON response.
 */
async function getJson<T = any>(url: string): Promise<T> {
  const response = await fetch(url, { headers: getHeaders() });
  return parseJson<T>(response);
}

/**
 * Execute POST request with JSON body and parse response.
 */
async function postJson<T = any>(url: string, body?: Json): Promise<T> {
  const response = await fetch(url, {
    method: 'POST',
    headers: getHeaders(),
    body: JSON.stringify(body || {}),
  });
  return parseJson<T>(response);
}

/**
 * Execute DELETE request and parse JSON response.
 */
async function deleteJson<T = any>(url: string): Promise<T> {
  const response = await fetch(url, {
    method: 'DELETE',
    headers: getHeaders(),
  });
  return parseJson<T>(response);
}

// ─────────────────────────────────────────────────────────────────────────────
// Dataset Operations
// ─────────────────────────────────────────────────────────────────────────────

const _refCache = new Map<string, DatasetRef>();

/**
 * Resolve dataset reference from name, with caching.
 */
export async function resolveDatasetRef(datasetName: string): Promise<DatasetRef> {
  const cached = _refCache.get(datasetName);
  if (cached) return cached;

  const result = await postJson<{
    success: boolean;
    dataset_id: string;
    dataset_name?: string;
  }>(`${API_BASE}/datasets/resolve`, { dataset_name: datasetName });

  if (!result.success || !result.dataset_id) {
    throw new Error(`Failed to resolve dataset "${datasetName}": ${JSON.stringify(result)}`);
  }

  const ref: DatasetRef = {
    appKey: datasetName,
    ragflowId: result.dataset_id,
    ragflowName: result.dataset_name ?? datasetName,
  };

  _refCache.set(datasetName, ref);
  return ref;
}

/**
 * Check RAGFlow health.
 */
export async function checkRAGFlowHealth(): Promise<{ status: string }> {
  return getJson<{ status: string }>(`${API_BASE}/health`);
}

/**
 * List datasets.
 */
export async function listDatasets(name?: string): Promise<RAGDataset[]> {
  const params = new URLSearchParams();
  if (name) params.set('name', name);
  const result = await getJson<{ datasets?: RAGDataset[] }>(
    `${API_BASE}/datasets?${params}`
  );
  return result.datasets ?? [];
}

/**
 * Create a new dataset.
 */
export async function createDataset(payload: {
  name: string;
  description?: string;
  chunk_method?: string;
}): Promise<RAGDataset> {
  return postJson<RAGDataset>(`${API_BASE}/datasets`, payload);
}

/**
 * Delete a dataset.
 */
export async function deleteDataset(datasetId: string): Promise<{ success: boolean }> {
  return deleteJson<{ success: boolean }>(`${API_BASE}/datasets/${datasetId}`);
}

// ─────────────────────────────────────────────────────────────────────────────
// Document Operations
// ─────────────────────────────────────────────────────────────────────────────

/**
 * List documents in a dataset with pagination.
 */
export async function listDocuments(
  datasetRef: DatasetRef,
  page = 1,
  pageSize = 30
): Promise<RAGDocument[]> {
  const params = new URLSearchParams();
  params.set('page', String(page));
  params.set('page_size', String(pageSize));

  const result = await getJson<{ documents?: RAGDocument[] }>(
    `${API_BASE}/datasets/${datasetRef.ragflowId}/documents?${params}`
  );
  return result.documents ?? [];
}

/**
 * Upload a file to a dataset.
 */
export async function uploadDocument(
  datasetId: string,
  file: { buffer: Buffer; originalname: string; mimetype: string },
  token: string // We need the JWT to get past Flask's @require_jwt
): Promise<any> {
  const form = new FormData();
  form.append('file', file.buffer, {
    filename: file.originalname,
    contentType: file.mimetype,
  });

  const url = `${AI_SERVICE_URL}/api/ragflow/datasets/${datasetId}/documents`;
  console.log(`[Node -> Flask] Uploading document to: ${url}`);

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      ...form.getHeaders(),
      'Authorization': `Bearer ${token}` // Forward the frontend's token to Flask
    },
    body: form,
  });

  return parseJson(response);
}

/**
 * Delete documents from a dataset.
 */
export async function deleteDocuments(
  datasetRef: DatasetRef,
  documentIds: string[],
  token: string
): Promise<{ success: boolean }> {
  
  // Hardcoding the full path structure to guarantee it hits Flask's blueprint perfectly
  const FLASK_URL = process.env.AI_SERVICE_URL || 'http://localhost:5001';
  const url = `${FLASK_URL}/api/ragflow/datasets/${datasetRef.ragflowId}/documents`;
  
  const response = await fetch(url, {
    method: 'DELETE', // Force explicit DELETE
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${token}`
    },
    body: JSON.stringify({ document_ids: documentIds })
  });

  const text = await response.text();
  if (!response.ok) {
    throw new Error(`Delete request failed with status ${response.status}: ${text}`);
  }

  return JSON.parse(text);
}

/**
 * Trigger chunking/parsing of documents.
 */
export async function parseDocuments(
  datasetId: string,
  documentIds: string[],
  token: string
): Promise<any> {
  const url = `${AI_SERVICE_URL}/api/ragflow/datasets/${datasetId}/chunks`;
  console.log(`[Node -> Flask] Triggering parse at: ${url} for docs:`, documentIds);
  
  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${token}`
    },
    body: JSON.stringify({ document_ids: documentIds }),
  });

  return parseJson(response);
}
/**
 * Extract document IDs from upload response in various formats.
 */
export function extractDocumentIds(uploadResponse: any): string[] {
  if (Array.isArray(uploadResponse?.document_ids)) {
    const ids = uploadResponse.document_ids.filter(Boolean);
    if (ids.length) return ids;
  }

  if (Array.isArray(uploadResponse?.data)) {
    const ids = uploadResponse.data
      .map((d: any) => d?.id ?? d?.document_id)
      .filter(Boolean);
    if (ids.length) return ids;
  }

  const single = uploadResponse?.document_id ?? uploadResponse?.id;
  if (single) return [single];

  if (Array.isArray(uploadResponse?.documents)) {
    const ids = uploadResponse.documents
      .map((d: any) => d?.id ?? d?.document_id)
      .filter(Boolean);
    if (ids.length) return ids;
  }

  throw new Error('No document IDs found in upload response');
}
// ─────────────────────────────────────────────────────────────────────────────
// Query & Retrieval
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Generate a module/lesson plan.
 */
export async function generateModule(payload: {
  class_name: string;
  subject: string;
  topic: string;
  teacher_id?: string;
  dataset_name?: string;
}): Promise<any> {
  return postJson(`${API_BASE}/module/generate`, payload);
}

/**
 * Ask a question (non-streaming).
 */
export async function askQuestion(payload: {
  question: string;
  teacher_id?: string;
  session_id?: string;
  chat_id?: string;
}): Promise<any> {
  return postJson(`${API_BASE}/query/ask`, { ...payload, stream: false });
}

/**
 * Retrieve chunks for a question (RAG retrieval).
 */
export async function retrieveQuestion(payload: {
  question: string;
  dataset_ids?: string[];
  dataset_name?: string;
  top_k?: number;
  similarity_threshold?: number;
}): Promise<{
  success: boolean;
  dataset_ids: string[];
  chunks: RAGChunk[];
  count: number;
}> {
  return postJson(`${API_BASE}/query/retrieve`, payload);
}

/**
 * Ask a question from a specific dataset.
 */
export async function askFromDataset(payload: {
  question: string;
  teacher_id?: string;
  session_id?: string;
  chat_id?: string;
  dataset_ids?: string[];
  dataset_name?: string;
  top_k?: number;
  similarity_threshold?: number;
}): Promise<any> {
  return postJson(`${API_BASE}/query/ask-from-dataset`, payload);
}

// ─────────────────────────────────────────────────────────────────────────────
// Streaming Query (SSE)
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Stream question response using Server-Sent Events (SSE).
 * Returns a Node.js Readable stream.
 */
export async function streamQuestion(payload: {
  question: string;
  teacher_id?: string;
  session_id?: string;
  chat_id?: string;
}): Promise<Readable> {
  const response = await fetch(`${API_BASE}/query/ask`, {
    method: 'POST',
    headers: getHeaders({ 'Content-Type': 'application/json' }),
    body: JSON.stringify({ ...payload, stream: true }),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(errorText || 'Streaming request failed');
  }

  if (!response.body) {
    throw new Error('No response body for streaming');
  }

  return Readable.fromWeb(response.body as any);
}

// ─────────────────────────────────────────────────────────────────────────────
// Resources
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Get resources.
 */
export async function getResources(
  teacherId?: string,
  competencyArea?: string
): Promise<RAGResource[]> {
  const params = new URLSearchParams();
  if (teacherId) params.set('teacher_id', teacherId);
  if (competencyArea) params.set('competency_area', competencyArea);

  const result = await getJson<{ resources?: RAGResource[] }>(
    `${API_BASE}/resources?${params}`
  );
  return result.resources ?? [];
}

/**
 * Get exemplary resources.
 */
export async function getExemplaryResources(): Promise<RAGResource[]> {
  const result = await getJson<{ resources?: RAGResource[] }>(
    `${API_BASE}/resources/exemplary`
  );
  return result.resources ?? [];
}

// ─────────────────────────────────────────────────────────────────────────────
// Session Management
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Create a chat session.
 */
export async function createSession(
  name?: string,
  chatId?: string
): Promise<any> {
  return postJson(`${API_BASE}/sessions`, {
    name: name ?? 'New Chat',
    chat_id: chatId,
  });
}

/**
 * List sessions.
 */
export async function listSessions(chatId?: string): Promise<any> {
  const params = new URLSearchParams();
  if (chatId) params.set('chat_id', chatId);
  return getJson(`${API_BASE}/sessions?${params}`);
}

/**
 * Delete a session.
 */
export async function deleteSession(
  sessionId: string,
  chatId?: string
): Promise<{ success: boolean }> {
  const params = new URLSearchParams();
  if (chatId) params.set('chat_id', chatId);
  const qs = params.toString();
  return deleteJson(`${API_BASE}/sessions/${sessionId}${qs ? `?${qs}` : ''}`);
}

File: Backend/ai-personalization/src/ragflow_client.py
"""
ragflow_client.py — RAGFlow /api/v1/* wrapper for the project

Provides low-level HTTP client functions for interacting with RAGFlow's REST API.
Handles authentication, request/response parsing, and error handling.
"""
# NOTE : Almost every function has been implemented but not in use (some of em) by ragflow_routes.
import os
from pathlib import Path
from typing import Dict, Any, List, Generator, Optional
import json

import requests
from dotenv import load_dotenv, find_dotenv

# Load environment variables
load_dotenv(find_dotenv(), override=False)

# Configuration
RAGFLOW_BASE_URL = os.getenv("RAGFLOW_BASE_URL", "http://localhost:80").rstrip("/")
RAGFLOW_API_KEY = os.getenv("RAGFLOW_API_KEY", "").strip()
RAGFLOW_CHAT_ID = os.getenv("RAGFLOW_CHAT_ID", "").strip()
RAGFLOW_TIMEOUT_SEC = int(os.getenv("RAGFLOW_TIMEOUT_SEC", "30"))

# ─────────────────────────────────────────────────────────────────────────────
# Helper Functions
# ─────────────────────────────────────────────────────────────────────────────


def _auth_headers(json_content: bool = True) -> Dict[str, str]:
    """Generate authorization headers for RAGFlow API requests."""
    if not RAGFLOW_API_KEY:
        raise ValueError("RAGFLOW_API_KEY is not set")

    headers = {"Authorization": f"Bearer {RAGFLOW_API_KEY}"}
    if json_content:
        headers["Content-Type"] = "application/json"
    return headers


def _url(path: str) -> str:
    """Construct full URL from RAGFlow base URL and path."""
    return f"{RAGFLOW_BASE_URL}{path}"


def _get(path: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """Execute GET request to RAGFlow API."""
    resp = requests.get(
        _url(path),
        headers=_auth_headers(json_content=True),
        params=params,
        timeout=RAGFLOW_TIMEOUT_SEC,
    )
    resp.raise_for_status()
    return resp.json()


def _post(
    path: str,
    json: Optional[Dict[str, Any]] = None,
    files: Optional[Dict[str, Any]] = None,
    stream: bool = False,
    timeout: Optional[int] = None,
):
    """Execute POST request to RAGFlow API."""
    headers = _auth_headers(json_content=(json is not None and files is None))
    resp = requests.post(
        _url(path),
        headers=headers,
        json=json,
        files=files,
        timeout=timeout or RAGFLOW_TIMEOUT_SEC,
        stream=stream,
    )
    resp.raise_for_status()
    return resp


def _delete(path: str, json: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """Execute DELETE request to RAGFlow API."""
    resp = requests.delete(
        _url(path),
        headers=_auth_headers(json_content=True),
        json=json,
        timeout=RAGFLOW_TIMEOUT_SEC,
    )
    resp.raise_for_status()
    return resp.json()

# ─────────────────────────────────────────────────────────────────────────────
# Chat Assistant Management Core Functions
# ─────────────────────────────────────────────────────────────────────────────

def _get_headers() -> Dict[str, str]:
    """Alias for _auth_headers — used by chat assistant management functions."""
    return _auth_headers(json_content=True)

def create_chat_assistant(name: str, dataset_ids: List[str], **kwargs) -> Dict[str, Any]:
    """Create a new RAGFlow chat assistant."""
    payload = {"name": name, "dataset_ids": dataset_ids, **kwargs}
    return _post("/api/v1/chats", json=payload)

def get_chat_assistant(chat_id: str) -> Dict[str, Any]:
    """Retrieve details for a specific chat assistant."""
    return _get(f"/api/v1/chats/{chat_id}")

def update_chat_assistant(chat_id: str, name: str, dataset_ids: List[str], **kwargs) -> Dict[str, Any]:
    """Full update (PUT) of a chat assistant's configuration."""
    payload = {"name": name, "dataset_ids": dataset_ids, **kwargs}
    return _put(f"/api/v1/chats/{chat_id}", json=payload)

def patch_chat_assistant(chat_id: str, payload: Dict[str, Any]) -> Dict[str, Any]:
    """Partial update (PATCH) of a chat assistant's configuration."""
    return _patch(f"/api/v1/chats/{chat_id}", json=payload)

def delete_chat_assistant(chat_id: str) -> Dict[str, Any]:
    """Delete a single chat assistant."""
    return _delete(f"/api/v1/chats/{chat_id}")

def list_chat_assistants(page: int = 1, page_size: int = 30, **kwargs) -> Dict[str, Any]:
    """List chat assistants with pagination filters."""
    params = {"page": page, "page_size": page_size, **kwargs}
    return _get("/api/v1/chats", params=params)

# ─────────────────────────────────────────────────────────────────────────────
# Health Check
# ─────────────────────────────────────────────────────────────────────────────


def health_check() -> bool:
    """Check if RAGFlow API is healthy."""
    try:
        resp = requests.get(_url("/api/v1/system/healthz"), timeout=5)
        resp.raise_for_status()
        data = resp.json()
        return data.get("status") == "ok"
    except Exception:
        return False


def health_detail() -> Dict[str, Any]:
    """Get detailed health information from RAGFlow API."""
    resp = requests.get(_url("/api/v1/system/healthz"), timeout=5)
    resp.raise_for_status()
    return resp.json()


# ─────────────────────────────────────────────────────────────────────────────
# Dataset Operations
# ─────────────────────────────────────────────────────────────────────────────


def list_datasets(name: str = "", page: int = 1, page_size: int = 30) -> List[Dict[str, Any]]:
    """List all datasets, optionally filtered by name."""
    params: Dict[str, Any] = {"page": page, "page_size": page_size}
    if name:
        params["name"] = name

    data = _get("/api/v1/datasets", params=params).get("data", {})
    if isinstance(data, list):
        return data
    if isinstance(data, dict):
        return data.get("datasets", data.get("list", [])) or []
    return []


def create_dataset(
    name: str, description: str = "", chunk_method: str = "naive"
) -> Dict[str, Any]:
    """Create a new dataset in RAGFlow."""
    resp = _post(
        "/api/v1/datasets",
        json={
            "name": name,
            "description": description,
            "chunk_method": chunk_method,
        },
    ).json()
    return resp


def get_or_create_dataset(
    name: str, description: str = "", chunk_method: str = "naive"
) -> Dict[str, Any]:
    """Get existing dataset or create if it doesn't exist."""
    existing = list_datasets(name=name)
    if existing:
        return existing[0]

    created = create_dataset(name, description, chunk_method)
    return created.get("data", created)


def delete_dataset(dataset_id: str) -> Dict[str, Any]:
    """Delete a dataset by ID."""
    return _delete("/api/v1/datasets", json={"ids": [dataset_id]})


# ─────────────────────────────────────────────────────────────────────────────
# Document Operations
# ─────────────────────────────────────────────────────────────────────────────


def list_documents(
    dataset_id: str, page: int = 1, page_size: int = 30
) -> List[Dict[str, Any]]:
    """List documents in a dataset."""
    data = _get(
        f"/api/v1/datasets/{dataset_id}/documents",
        params={"page": page, "page_size": page_size},
    ).get("data", {})

    if isinstance(data, dict):
        return data.get("docs", data.get("documents", [])) or []
    if isinstance(data, list):
        return data
    return []


def upload_document(dataset_id: str, file_path: str) -> Dict[str, Any]:
    """Upload a document file to a dataset."""
    file_name = Path(file_path).name
    with open(file_path, "rb") as f:
        resp = _post(
            f"/api/v1/datasets/{dataset_id}/documents",
            files={"file": (file_name, f, "application/pdf")},
            timeout=60,
        )
    return resp.json()


def parse_documents(dataset_id: str, document_ids: List[str]) -> Dict[str, Any]:
    """Trigger chunking/parsing of documents in a dataset."""
    return _post(
        f"/api/v1/datasets/{dataset_id}/chunks",
        json={"document_ids": document_ids},
    ).json()


def delete_documents(dataset_id: str, document_ids: List[str]) -> Dict[str, Any]:
    """Delete documents from a dataset."""
    return _delete(
        f"/api/v1/datasets/{dataset_id}/documents",
        json={"ids": document_ids},
    )


# ─────────────────────────────────────────────────────────────────────────────
# Retrieval Operations
# ─────────────────────────────────────────────────────────────────────────────


def retrieve_chunks(
    question: str,
    dataset_ids: List[str],
    top_k: int = 6,
    similarity_threshold: float = 0.2,
) -> List[Dict[str, Any]]:
    """Retrieve relevant chunks from datasets based on a question."""
    resp = _post(
        "/api/v1/retrieval",
        json={
            "question": question,
            "dataset_ids": dataset_ids,
            "top_k": top_k,
            "similarity_threshold": similarity_threshold,
        },
    ).json()

    data = resp.get("data", {})
    if isinstance(data, dict):
        return data.get("chunks", [])
    return []


# ─────────────────────────────────────────────────────────────────────────────
# Chat & Completion Operations
# ─────────────────────────────────────────────────────────────────────────────


def _resolve_chat_id(chat_id: str = "") -> str:
    """Resolve chat ID, falling back to environment default if needed."""
    cid = (chat_id or RAGFLOW_CHAT_ID).strip()
    if not cid:
        raise ValueError("No RAGFLOW_CHAT_ID configured")
    return cid


def chat_completion(
    question: str, chat_id: str = "", session_id: Optional[str] = None
) -> Dict[str, Any]:
    cid = _resolve_chat_id(chat_id)

    payload:Dict[str,Any] = {
        "model": "model",
        "messages": [{"role": "user", "content": question}],
        "stream": False,
        "extra_body": {
            "reference": True,
            "reference_metadata": {
                "include": True
            }
        }
    }

    return _post(
        f"/api/v1/openai/{cid}/chat/completions",
        json=payload,
        timeout=90,
    ).json()


def chat_completion_stream(
    question: str,
    chat_id: str = "",
    session_id: Optional[str] = None,
    **kwargs,  
) -> Generator[str, None, None]:
    """Get a chat completion response with streaming."""
    cid = _resolve_chat_id(chat_id)

    payload: Dict[str, Any] = {
        "model": "model",
        "messages": [{"role": "user", "content": question}],
        "stream": True,
        "extra_body": {
            "reference": True,
            "reference_metadata": {
                "include": True
            }
        }
    }

    if session_id:
        payload["session_id"] = session_id

    if "dataset_ids" in kwargs and kwargs["dataset_ids"]:
        payload["dataset_ids"] = kwargs["dataset_ids"]

    resp = _post(
        f"/api/v1/openai/{cid}/chat/completions",
        json=payload,
        stream=True,
        timeout=120,
    )
    references = []

    for chunk in resp.iter_lines():
        if not chunk:
            continue
        
        chunk_str = chunk.decode("utf-8")
        if not chunk_str.startswith("data:"):
            continue
        
        data_str = chunk_str[5:].strip()
        if data_str == "[DONE]":
            break
        
        try:
            data = json.loads(data_str)
            
            delta = data.get("choices", [{}])[0].get("delta") or {}
            ref = delta.get("reference") or []
            if ref:
                references = ref
            
            yield f"data: {data_str}\n\n"
        except json.JSONDecodeError:
            continue
    
    yield f'event: metadata\ndata: {json.dumps({"references": references})}\n\n'

def chat_completion_stream_stateless(messages: List[Dict[str, Any]], chat_id: str) -> Generator[str, None, None]:
    """Stateless streaming: caller passes the full message history array."""
    cid = (chat_id or RAGFLOW_CHAT_ID).strip()
    url = f"{RAGFLOW_BASE_URL}/api/v1/openai/{cid}/chat/completions"
    payload = {"model": "model", "messages": messages, "stream": True}
    resp = requests.post(url, json=payload, headers=_auth_headers(), stream=True, timeout=120)
    resp.raise_for_status()
    for line in resp.iter_lines():
        if line:
            yield line.decode("utf-8")

def chat_completion_stream_stateful(
    question: str,
    session_id: str,
    chat_id: str
) -> Generator[str, None, None]:
    """Stateful streaming: RAGFlow tracks history server-side via session_id."""
    cid = (chat_id or RAGFLOW_CHAT_ID).strip()
    url = f"{RAGFLOW_BASE_URL}/api/v1/chats/{cid}/completions"
    payload = {
        "question": question,
        "session_id": session_id,
        "stream": True
    }
    resp = requests.post(
        url,
        json=payload,
        headers=_auth_headers(),
        stream=True,
        timeout=120
    )
    resp.raise_for_status()
    for line in resp.iter_lines():
        if line:
            yield line.decode("utf-8")

# ─────────────────────────────────────────────────────────────────────────────
# Session Operations
# ─────────────────────────────────────────────────────────────────────────────


def create_session(chat_id: str = "", name: str = "Guru-Sikshan Session") -> Dict[str, Any]:
    """Create a new chat session."""
    cid = _resolve_chat_id(chat_id)
    print("[ragflow_client] session url:", _url(f"/api/v1/chats/{cid}/sessions"))
    print("[ragflow_client] chat id:", cid)
    cid = _resolve_chat_id(chat_id)
    return _post(
        f"/api/v1/chats/{cid}/sessions",
        json={"name": name},
    ).json()


def list_sessions(chat_id: str = "") -> List[Dict[str, Any]]:
    """List all sessions for a chat."""
    cid = _resolve_chat_id(chat_id)
    data = _get(f"/api/v1/chats/{cid}/sessions").get("data", {})
    if isinstance(data, list):
        return data
    if isinstance(data, dict):
        return data.get("sessions", data.get("list", [])) or []
    return []


def delete_session(chat_id: str, session_id: str) -> Dict[str, Any]:
    """Delete a session."""
    cid = _resolve_chat_id(chat_id)
    return _delete(
        f"/api/v1/chats/{cid}/sessions",
        json={"ids": [session_id]},
    )

def get_chat_session(chat_id: str, session_id: str) -> Dict[str, Any]:
    """
    Retrieve a specific chat session's history and messages.
    """
    return _get(f"/api/v1/chats/{chat_id}/sessions/{session_id}")

